Core/Core/Planner/PlannerFilterViewController.swift

```diff
@@ -0,0 +1,180 @@
+        tableView.reloadData()
+    }
+
+    @objc func refresh(_ control: CircleRefreshControl) {
+        courses.refresh(force: true)
+    }
+
+    func toggleCourse(_ course: Course) throws {
+        if planner?.selectedCourses.contains(course) == true {
+            planner?.selectedCourses.remove(course)
+        } else {
+            planner?.selectedCourses.insert(course)
+        }
+        try env.database.viewContext.save()
+    }
+}
+
+extension PlannerFilterViewController: UITableViewDataSource {
+    public func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
+
+    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        var count = courses.count
+        if courses.hasNextPage {
+            count += 1
+        }
+        return count
+    }
+
+    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if courses.hasNextPage && indexPath.row == courses.count {
+            courses.getNextPage()
+            return LoadingCell(style: .default, reuseIdentifier: nil)
+        }
+        let cell = tableView.dequeue(for: indexPath) as PlannerFilterCell
+        let course = courses[indexPath]
+        cell.courseNameLabel.text = course?.courseCode
+        cell.isChecked = course.flatMap { planner?.selectedCourses.contains($0) } ?? false
+        return cell
+    }
+
+    public func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard section == 0 else { return nil }
+        let header = tableView.dequeueHeaderFooter(SectionHeaderView.self)
+        header.titleLabel?.text = NSLocalizedString("Courses", comment: "")
+        return header
+    }
+}
+
+extension PlannerFilterViewController: UITableViewDelegate {
+    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+        guard let course = courses[indexPath] else { return }
+        do {
+            try toggleCourse(course)
+        } catch {
+            showError(error)
+        }
+    }
+}
+
+class PlannerFilterCell: UITableViewCell {
+    @IBOutlet weak var checkboxView: UIView!
+    @IBOutlet weak var courseNameLabel: UILabel!
+    @IBOutlet weak var checkmark: UIImageView!
+
+    var isChecked: Bool = false {
+        didSet {
+            checkboxView.layer.borderWidth = isChecked ? 0 : 1
+            checkmark.isHidden = !isChecked
+        }
+    }
+
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        checkboxView.backgroundColor = .clear
+        checkboxView.layer.cornerRadius = 3
+        checkboxView.layer.borderWidth = 1
+        checkboxView.layer.borderColor = UIColor.named(.borderDark).cgColor
+        checkmark.tintColor = .named(.backgroundInfo)
+        selectionStyle = .none
+    }
+}
+
+class LoadingCell: UITableViewCell {
```

Comment:
I've seen the pattern of LoadingCells before. Do we not have a reusable cell for this yet?

Reply:
Oh yeah I meant to break this out into its own thing so that we can properly re-use it. Thanks for the reminder.